src/Microsoft.OData.Core/Json/JsonReader.cs

```diff
@@ -1316,25 +2262,35 @@ private bool ReadInput()
             Debug.Assert(
                 this.storedCharacterCount < this.characterBuffer.Length,
                 "We should have more room in the buffer by now.");
+        }
 
-            // Read more characters from the input.
-            // Use the Read method which returns any character as soon as it's available
-            // we don't want to wait for the entire buffer to fill if the input doesn't have
-            // the characters ready.
-            int readCount = this.reader.Read(
-                this.characterBuffer,
-                this.storedCharacterCount,
-                this.characterBuffer.Length - this.storedCharacterCount);
-
-            if (readCount == 0)
+        /// <summary>
+        /// Parses a unicode hex value.
+        /// </summary>
+        /// <returns>32-bit signed integer equivalent.</returns>
+        private int ParseUnicodeHexValue()
+        {
+            string unicodeHexValue = this.ConsumeTokenToString(4);
+            int characterValue;
+            if (!Int32.TryParse(unicodeHexValue, NumberStyles.HexNumber, CultureInfo.InvariantCulture, out characterValue))
             {
-                // No more characters available, end of input.
-                this.endOfInputReached = true;
-                return false;
+                throw JsonReaderExtensions.CreateException(Strings.JsonReader_UnrecognizedEscapeSequence("\\u" + unicodeHexValue));
             }
 
-            this.storedCharacterCount += readCount;
-            return true;
+            return characterValue;
+        }
+
+        /// <summary>
+        /// Determines if a given character is allowed in a JSON property name.
+        /// </summary>
+        /// <param name="character">The character to check.</param>
+        /// <returns>true if the character is allowed in a JSON property name; otherwise false.</returns>
+        private bool IsCharacterAllowedInPropertyName(char character)
+        {
+            return character == '_' ||

```

Comment:
I'd be interested to see a benchmark comparing this with an equivalent `switch` statement

Reply:
@corranrogue9 The `switch` equivalent would look something like this:
```csharp
switch (character)
{
    case '_':
    case '$':
        return true;
    default:
        if (Char.IsLetterOrDigit(character) || (this.allowAnnotations && (character == '.' || character == '@')))
            return true;
        else
            return false;
}
```
I'm not sure I like this alternative better. However, one micro-optimization that is easy to apply is making `Char.IsLetterOrDigit(character)` the first condition to check given that letters and digits are the most common characters in a property name.

Reply:
I would definitely like to see that benchmark if you do it. I think `switch` is supposed to be O(~1), so I think running any code that has branching logic before the `switch` would make it overall slower. 

As for this feedback, you can take it or leave it, just a suggestion

Reply:
@corranrogue9 I grabbed the csdl for a particular workload that receives relatively high traffic and set up 2 benchmarks - the first benchmark calling a variant of `IsCharacterAllowedInPropertyName` method that uses a `switch` statement and the second benchmark calling a variant using the code in my PR. The idea being to verify that the characters that make up the properties in the different entity and complex types are allowed. Here are the observed benchmark results after 5 separate runs:
```
|         Method |     Mean |    Error |   StdDev |
|--------------- |---------:|---------:|---------:|
| NotUsingSwitch | 77.49 us | 1.244 us | 1.103 us |
|    UsingSwitch | 76.78 us | 1.060 us | 1.041 us |

|         Method |     Mean |    Error |   StdDev |
|--------------- |---------:|---------:|---------:|
| NotUsingSwitch | 59.80 us | 0.191 us | 0.159 us |
|    UsingSwitch | 79.53 us | 1.561 us | 1.603 us |

|         Method |     Mean |    Error |   StdDev |
|--------------- |---------:|---------:|---------:|
| NotUsingSwitch | 77.14 us | 1.537 us | 3.106 us |
|    UsingSwitch | 80.51 us | 1.577 us | 3.112 us |

|         Method |     Mean |    Error |   StdDev |
|--------------- |---------:|---------:|---------:|
| NotUsingSwitch | 95.47 us | 1.397 us | 1.239 us |
|    UsingSwitch | 93.38 us | 1.785 us | 2.321 us |

|         Method |     Mean |    Error |   StdDev |
|--------------- |---------:|---------:|---------:|
| NotUsingSwitch | 61.46 us | 0.993 us | 0.829 us |
|    UsingSwitch | 80.51 us | 1.394 us | 3.029 us |
```